optimizer/plan/planbuilder.go

```diff
@@ -613,3 +615,85 @@ func (se *subqueryVisitor) Enter(in ast.Node) (out ast.Node, skipChildren bool)
 func (se *subqueryVisitor) Leave(in ast.Node) (out ast.Node, ok bool) {
 	return in, true
 }
+
+func (b *planBuilder) buildUpdate(update *ast.UpdateStmt) Plan {
+	sel := &ast.SelectStmt{From: update.TableRefs, Where: update.Where, OrderBy: update.Order, Limit: update.Limit}
+	p := b.buildFrom(sel)
+	if sel.OrderBy != nil && !matchOrder(p, sel.OrderBy.Items) {
+		p = b.buildSort(p, sel.OrderBy.Items)
+		if b.err != nil {
+			return nil
+		}
+	}
+	if sel.Limit != nil {
+		p = b.buildLimit(p, sel.Limit)
+		if b.err != nil {
+			return nil
+		}
+	}
+	orderedList := b.buildUpdateLists(update.List, p.Fields())
+	if b.err != nil {
+		return nil
+	}
+	return &Update{OrderedList: orderedList, SelectPlan: p}
+}
+
+func (b *planBuilder) buildUpdateLists(list []*ast.Assignment, fields []*ast.ResultField) []*ast.Assignment {
+	newList := make([]*ast.Assignment, len(fields))
+	for _, assign := range list {
+		idx, err := columnIndexInFields(assign.Column, fields)
+		if err != nil {
+			b.err = errors.Trace(err)
+			return nil
+		}
+		newList[idx] = assign
+	}
+	return newList
+}
+
+func columnIndexInFields(cn *ast.ColumnName, fields []*ast.ResultField) (int, error) {
```

Comment:
I suggest that it is better to use columnOffset instead of columnIndex, for index is a little confused.